pkg/controllers/status/cluster_condition_cache_test.go

```diff
@@ -0,0 +1,119 @@
+package status
+
+import (
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	clusterv1alpha1 "github.com/karmada-io/karmada/pkg/apis/cluster/v1alpha1"
+)
+
+func TestThresholdAdjustedReadyCondition(t *testing.T) {
+	tests := []struct {
+		online        bool
+		healthy       bool
+		expectedReady metav1.ConditionStatus
+	}{
+		{
+			// cluster is healthy now
+			online:        true,
+			healthy:       true,
+			expectedReady: metav1.ConditionTrue,
+		},
+		{
+			// cluster becomes unhealthy
+			online:        true,
+			healthy:       false,
+			expectedReady: metav1.ConditionTrue,
+		},
+		{
+			// the first retry
+			online:        false,
+			healthy:       true,
+			expectedReady: metav1.ConditionTrue,
+		},
+		{
+			// the second retry
+			online:        false,
+			healthy:       false,
+			expectedReady: metav1.ConditionTrue,
+		},
+		{
+			// the third retry
+			// cluster is still unhealthy after the threshold, set cluster status to not ready
+			online:        false,
+			healthy:       false,
+			expectedReady: metav1.ConditionFalse,
+		},
+		{
+			// cluster is still unhealthy after the threshold, set cluster status to not ready
+			online:        false,
+			healthy:       false,
+			expectedReady: metav1.ConditionFalse,
+		},
+		{
+			// cluster recoverd
+			online:        true,
+			healthy:       true,
+			expectedReady: metav1.ConditionTrue,
+		},
+		{
+			// cluster becomes unhealthy
+			online:        false,
+			healthy:       false,
+			expectedReady: metav1.ConditionTrue,
+		},
+		{
+			// cluster recoverd
+			online:        true,
+			healthy:       true,
+			expectedReady: metav1.ConditionTrue,
+		},
+		{
+			// cluster becomes unhealthy
+			online:        false,
+			healthy:       false,
+			expectedReady: metav1.ConditionTrue,
+		},
+		{
+			// the first retry
+			online:        false,
+			healthy:       false,
+			expectedReady: metav1.ConditionTrue,
+		},
+		{
+			// the second retry
+			online:        false,
+			healthy:       false,
+			expectedReady: metav1.ConditionTrue,
+		},
+		{
+			// the third retry
+			// cluster is still unhealthy after the threshold, set cluster status to not ready
+			online:        false,
+			healthy:       false,
+			expectedReady: metav1.ConditionFalse,
+		},
```

Comment:
Seems these tests interact with each other?

Reply:
Yes thery are observed statuses ordered by time, because `thresholdAdjustedReadyCondition` is stateful, it depends on previous statuses

Reply:
In this test, we are testing `thresholdAdjustedReadyCondition`, we should set up `cache`, `current condition`, and `observed condition` for each case. They should not depend on each other.

Reply:
That would be kind of tricky..How to set up the the `cache`?(`clusterDataMap` and `thresholdStartTime`)

Reply:
We are testing `clusterConditionStore.thresholdAdjustedReadyCondition` where `clusterConditionStore` is stateful